### ISSUE TITLE
setting different optimization parameters for different layers / different epochs

### ISSUE THREAD

**Issue by denizyuret** [denizyuret/Knet.jl#449]
Current design:
* parameters have an opt field specifying the optimization algorithm/options.
* by default these are initialized as `nothing`.
* call to a minimizer (sgd, adam etc.) sets the opt fields to specified optimizer if empty.
* however, they do not touch the opt field if already set.
* this gives the user the freedom to set the optimization algorithm/options differently for different layers if needed.
* however if the user wants to say reduce the learning rate for the tenth epoch and calls sgd with a new learning rate, this is silently ignored right now because sgd sees the opt fields already set and does not change them, using the initially specified learning rates.

Since reducing learning rate is a frequent practice the current design needs to change.

**Comment by ngphuoc** [comment]
Currently, I'd like to change the learning rate. Is creating a new model and copying the old params over for training with the new learning rate the only approach?

**Comment by ekinakyurek** [comment]
You don’t need to copy them. This is the way I apply learning rate decay:

https://github.com/ekinakyurek/Morse.jl/blob/a9190261d480b56a9f97c66ece8193602106b315/src/util.jl#L84

**Comment by ngphuoc** [comment]
Thanks! That would be simple enough given I am having a callback at every specified epochs/step in my train! function.

**Comment by s271** [comment]
This is related to
[https://github.com/denizyuret/Knet.jl/issues/438](update! seems not working on truncted weights/gradients)

**Comment by davidssmith** [comment]
It would be ideal to be able to pass a function as the `lr` parameter with the form `lr = f(epoch)`.

Alternatively, the model `darknet` uses where a learning rate decay can be specified by a shape (e.g. exponential, polynomial, etc.) and some shape parameters (e.g. power law index) could work.  Here are some examples:
```
learning_rate=0.01
policy=poly
power=4
```
```
learning_rate=0.1
policy=steps
steps=1000,1500
scales=.1,.1
```
